src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -0,0 +1,382 @@
+import org.apache.cassandra.utils.bytecomparable.ByteSourceInverse;
+
+/**
+ * This is an in-memory index using the {@link InMemoryTrie} to store a {@link ByteComparable}
+ * representation of the indexed values. Data is stored on-heap or off-heap and follows the
+ * settings of the {@link TrieMemtable} to determine where.
+ *
+ *
+ */
+public class TrieMemoryIndex
+{
+    private static final Logger logger = LoggerFactory.getLogger(TrieMemoryIndex.class);
+    private static final int MINIMUM_QUEUE_SIZE = 128;
+    private static final int MAX_RECURSIVE_KEY_LENGTH = 128;
+
+    private final IndexContext indexContext;
+    private final InMemoryTrie<PrimaryKeys> data;
+    private final PrimaryKeysReducer primaryKeysReducer;
+    private final AbstractAnalyzer.AnalyzerFactory analyzerFactory;
+    private final AbstractType<?> validator;
+    private final boolean isLiteral;
+    private final Object writeLock = new Object();
+
+    private ByteBuffer minTerm;
+    private ByteBuffer maxTerm;
+
+    private static final FastThreadLocal<Integer> lastQueueSize = new FastThreadLocal<Integer>()
+    {
+        protected Integer initialValue()
+        {
+            return MINIMUM_QUEUE_SIZE;
+        }
+    };
+
+
+    public TrieMemoryIndex(IndexContext indexContext)
+    {
+        this.indexContext = indexContext;
+        this.data = new InMemoryTrie<>(TrieMemtable.BUFFER_TYPE);
+        this.primaryKeysReducer = new PrimaryKeysReducer();
+        // The use of the analyzer is within a synchronized block so can be considered thread-safe
+        this.analyzerFactory = indexContext.getIndexAnalyzerFactory();
+        this.validator = indexContext.getValidator();
+        this.isLiteral = TypeUtil.isLiteral(validator);
+    }
+
+    /**
+     * Adds a index value to the in-memory index
+     *
+     * @param key partition key for the indexed value
+     * @param clustering clustering for the indexed value
+     * @param value indexed value
+     * @return amount of heap allocated by the new value
+     */
+    public long add(DecoratedKey key, Clustering<?> clustering, ByteBuffer value)
+    {
+        synchronized (writeLock)
+        {
+            AbstractAnalyzer analyzer = analyzerFactory.create();
+            try
+            {
+                value = TypeUtil.encode(value, validator);
+                analyzer.reset(value.duplicate());
+                final PrimaryKey primaryKey = indexContext.keyFactory().create(key, clustering);
+                final long initialSizeOnHeap = data.sizeOnHeap();
+                final long initialSizeOffHeap = data.sizeOffHeap();
+                final long reducerHeapSize = primaryKeysReducer.heapAllocations();
+
+                while (analyzer.hasNext())
+                {
+                    final ByteBuffer term = analyzer.next();
+
+                    setMinMaxTerm(term.duplicate());
+
+                    final ByteComparable encodedTerm = encode(term.duplicate());
```

Comment:
nit/naming: `comparableTerm`

There's a similar thing w/ the `encode()` method here vs. the one in `TypeUtil`. One does our own thing to convert certain types to a fixed width format, and the other does the normal `asComparableBytes()` conversion. Perhaps this `encode()` could be like `asComparable()` or `asComparableBytes()`...